packages/wrangler/src/api/startDevWorker/RemoteRuntimeController.ts

```diff
@@ -1,30 +1,303 @@
+				host: config.dev.origin?.hostname,
+				routes,
+				sendMetrics: config.sendMetrics,
+			});
+
+			const bindings = (
+				await convertBindingsToCfWorkerInitBindings(config.bindings)
+			).bindings;
+
+			const token = await this.#previewToken({
+				bundle,
+				modules: bundle.modules,
+				accountId: auth.accountId,
+				name: config.name,
+				legacyEnv: config.legacyEnv,
+				env: config.env,
+				isWorkersSite: config.site !== undefined,
+				assetPaths: config.site?.path
+					? {
+							baseDirectory: config.site.path,
+							assetDirectory: "",
+							excludePatterns: config.site.exclude ?? [],
+							includePatterns: config.site.include ?? [],
+						}
+					: undefined,
+				format: bundle.entry.format,
+				// TODO: Remove this passthrough
+				bindings: config._bindings ? config._bindings : bindings,
+				compatibilityDate: config.compatibilityDate,
+				compatibilityFlags: config.compatibilityFlags,
+				usageModel: config.usageModel,
+				routes,
+			});
+
+			// If we received a new `bundleComplete` event before we were able to
+			// dispatch a `reloadComplete` for this bundle, ignore this bundle.
+			// If `token` is undefined, we've surfaced a relevant error to the user above, so ignore this bundle
+			if (id !== this.#currentBundleId || !token) {
+				return;
+			}
+
+			const accessToken = await getAccessToken(token.host);
+
+			this.emitReloadCompleteEvent({
+				type: "reloadComplete",
+				bundle,
+				config,
+				proxyData: {
+					userWorkerUrl: {
+						protocol: "https:",
+						hostname: token.host,
+						port: "443",
+					},
+					userWorkerInspectorUrl: {
+						protocol: token.inspectorUrl.protocol,
+						hostname: token.inspectorUrl.hostname,
+						port: token.inspectorUrl.port.toString(),
+						pathname: token.inspectorUrl.pathname,
+					},
+					userWorkerInnerUrlOverrides: {
+						hostname: token.host,
+					},
+					headers: {
+						"cf-workers-preview-token": token.value,
+						...(accessToken
+							? { Cookie: `CF_Authorization=${accessToken}` }
+							: {}),
+					},
+					liveReload: config.dev.liveReload,
+					proxyLogsToController: true,
+					internalDurableObjects: [],
+					entrypointAddresses: {},
+				},
+			});
+		} catch (error) {
+			this.emitErrorEvent({
+				type: "error",
+				reason: "Error reloading remote server",
+				cause: castErrorCause(error),
+				source: "RemoteRuntimeController",
+				data: undefined,
+			});
+		}
+	}
+
 	// ******************
 	//   Event Handlers
 	// ******************
 
 	onBundleStart(_: BundleStartEvent) {
-		notImplemented(this.onBundleStart.name, this.constructor.name);
+		// Abort any previous operations when a new bundle is started
+		this.#abortController.abort();
+		this.#abortController = new AbortController();
 	}
-	onBundleComplete(_: BundleCompleteEvent) {
-		notImplemented(this.onBundleComplete.name, this.constructor.name);
+	onBundleComplete(ev: BundleCompleteEvent) {
+		const id = ++this.#currentBundleId;
```

Comment:
I'd prefer we didn't have this incrementing id stuff – we can accomplish the same thing with identity checks of `bundle` and `config` (see `latestConfig` in ProxyController)

Reply:
Let's address in a followup—for now this mimics the patterns in the local runtime controller